custom_components/airthings_wave/airthings.py

```diff
@@ -192,152 +206,159 @@ def __init__(self, scan_interval, mac=None):
+                                self._dev = None
+                    self.devices[mac] = device    
+                except:
+                    _LOGGER.exception("Error getting device info.")
+                await self.disconnect()
+            else:
+                _LOGGER.error("Not getting device info because failed to connect to device.")
         return self.devices
 
-    def get_sensors(self):
+    async def get_sensors(self):
         self.sensors = {}
         for mac in self.airthing_devices:
-            self.connect(mac)
-            if self._dev is not None:
-                try:
-                    characteristics = self._dev.getCharacteristics()
-                    sensor_characteristics =  []
-                    for characteristic in characteristics:
+            await self.connect(mac)
+            if self._dev is not None and self._dev.is_connected:
+                sensor_characteristics =  []
+                svcs = await self._dev.get_services()
+                for service in svcs:
+                    for characteristic in service.characteristics:
                         _LOGGER.debug(characteristic)
                         if characteristic.uuid in sensors_characteristics_uuid_str:
                             sensor_characteristics.append(characteristic)
-                    self.sensors[mac] = sensor_characteristics
-                except btle.BTLEDisconnectError:
-                        _LOGGER.exception("Disconnected")
-                        self._dev = None
-            self.disconnect()
+                self.sensors[mac] = sensor_characteristics
+            await self.disconnect()                            
         return self.sensors
 
-    def get_sensor_data(self):
+    async def get_sensor_data(self):
         if time.monotonic() - self.last_scan > self.scan_interval or self.last_scan == -1:
             self.last_scan = time.monotonic()
             for mac, characteristics in self.sensors.items():
-                self.connect(mac)
-                if self._dev is not None:
+                await self.connect(mac)
+                if self._dev is not None and self._dev.is_connected:
                     try:
                         for characteristic in characteristics:
                             sensor_data = None
                             if str(characteristic.uuid) in sensor_decoders:
-                                char = self._dev.getCharacteristics(uuid=characteristic.uuid)[0]
-                                data = char.read()
+                                data = await self._dev.read_gatt_char(characteristic.uuid)
                                 sensor_data = sensor_decoders[str(characteristic.uuid)].decode_data(data)
                                 _LOGGER.debug("{} Got sensordata {}".format(mac, sensor_data))
-                                
+                            
+                            # ToDo: Is there a more elegant way to handle the below?
```

Comment:
Use an asyncio [synchronization primitive](https://docs.python.org/3/library/asyncio-sync.html) instead of polling.

Reply:
I fully agree with the polling. I introduced an asyncio.Event. As I understand the Event it is exactly what we need here. Can you have another look?

Reply:
While unlikely, the challenge with event.wait() is that there is no timeout value and it could block indefinitely if something happens, like a bluetooth disconnect, at the wrong time. I suggest using wait_for instead with a 1 second timeout, which I have tested and works for me:

```
# Wait for up to one second to see if a callblack comes in.
try:
    await asyncio.wait_for(self._event.wait(), 1)
except asyncio.TimeoutError:
    _LOGGER.warn("Timeout getting command data.")
if self._command_data is not None:
    sensor_data = command_decoders[str(characteristic.uuid)].decode_data(self._command_data)
    self._command_data = None
# Stop notification handler
```

You can see the full code [here](https://github.com/mjmccans/airthings-mqtt-ha/blob/dev/src/airthings.py).